lessons/specifics/debugging.md

```diff
@@ -0,0 +1,190 @@
+ 
+```elixir
+defp deps do
+  [{:dialyxir, "~> 0.4", only: [:dev]}]
+end
+```
+
+Then we call:
+
+```shell
+$ mix deps.get
+...
+$ mix deps.compile
+```
+
+First command download and install Dialyxir. You could be asked for installing Hex. Second compile Dialyxir app. If you want to install Dialyxir as global app, please read [documentation](https://github.com/jeremyjh/dialyxir#installation).
+
+Last step is to run Dialyzer to rebuild PLT. You need to do this every time after installation of new version of Erlang or Elixir. Thanks to this Dialyzer will not be try to analyse standard library every time that you try to use it. It takes few minutes.
+
+```shell
+$ mix dialyzer.plt
+Starting PLT Core Build ... this will take awhile
+dialyzer --build_plt --output_plt /.dialyxir_core_18_1.3.2.plt --apps erts kernel stdlib crypto public_key -r /Elixir/lib/elixir/../eex/ebin /Elixir/lib/elixir/../elixir/ebin /Elixir/lib/elixir/../ex_unit/ebin /Elixir/lib/elixir/../iex/ebin /Elixir/lib/elixir/../logger/ebin /Elixir/lib/elixir/../mix/ebin
+  Creating PLT /.dialyxir_core_18_1.3.2.plt ...
+...
+ done in 5m14.67s
+done (warnings were emitted)
+```
+
+### Static analysis of code
+ 
+Now, when we have Dialyxir on board we could run it:
+ 
+```shell
+$ mix dialyzer
+...
+examples.ex:3: Invalid type specification for function 'Elixir.Examples':sum_times/1. The success typing is (_) -> number()
+...
+```
+
+Communicate is clear. Return type of our function `sum_times/1` is different than declared. That because `Enum.sum` return `number` not `integer` and effective return type of `sum_times/1` is `number`. Because `number` is not `integer` so we have error. How to fix it? We could add `round` function that change `number` to `integer`:
+
+```elixir
+@spec sum_times(integer) :: integer
+def sum_times(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+    |> round
+end
+```
+
+Finally:
+
+```shell
+$ mix dialyzer
+...
+  Proceeding with analysis... done in 0m0.95s
+done (passed successfully)
+```
+
+Using specifications with tools to static code analysis helps us to make code that is self tested and contains less bugs.  
+
+## Debugging
+
+Sometimes static analysis of code is not enough. We need to know what is a real flow of code, because we would like find bugs. Simplest way is to put in code `IO.puts` to track values and code flow, but this technique is pretty primitive and limited. In Elixir we have could use erlang debugger to debug code. 
+
+Let’s look at a basic module:
+
+```elixir
+defmodule Example do
+
+  def cpu_burns(a, b, c) do
+    x = a * 2
+    y = b * 3
+    z = c * 5
+    
+    x + y + z
+  end
+
+end
+```
+
+Then run `iex`:
+ 
+```bash
+$ iex -S mix
+```
+
+And run debugger:
+
+```elixir
+iex > :debugger.start()
+{:ok, #PID<0.307.0>}
+```
+
+The `:debugger` is Erlang module that give us access to debugger. We could use function `start/1` to configure debugger:

```

Comment:
The `:debugger` is Erlang module that give us access to debugger. We could use function `start/1` to configure debugger:

-->

The `:debugger` is an Erlang module that give us access to a debugger built into the language. We could use its function `start/1` to configure it: